Menu drilldown

kmcfaul opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Follows up PR #5321. Closes #5024 and begins to address #5333 (+breadcrumbs, not yet on breadcrumb dropdowns)

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-5485.surge.sh

kmcfaul commented:
@jschuler I re-added the other menu demos, is that fine or was there a reason they were removed?

jschuler commented:
@kmcfaul thanks for adding them back in, I had just removed them temporarily to work on this feature :)

mcoker reviewed (approved):
LGTM!

kmcfaul commented:
Fixed the animation bug for sub-level drilldowns.
Also added the first round for drilldown breadcrumbs in a separate demo. Doesn't cover breadcrumb dropdowns yet.

Went with using our current Breadcrumb components over making a container "DrilldownMenuHeader" component. This way gives the user finer control over what the Breadcrumb/BreadcrumbItem should do and look like (opposed to passing in an array of data to an aggregate component) and later down the line will let the user control where in the chain the breadcrumb dropdown should be. 

I did have to update Breadcrumb to get the structure we use in core for the link styling, if using the `component` prop to do this interferes with anything else I can take it back out and make a wrapper component in Menu.

@mcoker I do have a question about using the link version of BreadcrumbHeading, which is in the core example. Is it ok just to pass a string for the drilldown? An href of "#" will reset the whole drilldown (because the page reloads).

mcarrano reviewed (commented):
This is looking great @kmcfaul . Love the animation effects. My only question/ask if if we can expose the label used for the Root node via a prop. The consumer might want to replace this with something more germane to their use case. @doruskova can you also take a look when you have a chance?

kmcfaul commented:
@mcarrano The text in the breadcrumbs is defined on the app-side so it should be no problem to change it. Keyboard interaction is not in yet beyond any default handling by the browser, we will likely need a follow up issue for it.

redallen commented on packages/react-core/src/components/Menu/examples/Menu.md:
```diff
@@ -692,3 +700,535 @@ class MenuOptionMultiSelect extends React.Component {
+      >
+        <MenuContent
+          getHeight={height => this.setHeight('rootMenu', height)}
+          style={{ '--pf-c-menu__content--Height': `${menuHeights[activeMenu]}px` }}
```
If we want setting the menu height based off the drilled down MenuItem to be the default behavior can we store that state in MenuContent instead?

redallen commented on packages/react-core/src/components/Menu/examples/Menu.md:
```diff
@@ -692,3 +700,535 @@ class MenuOptionMultiSelect extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      activeItem: null,
```
It doesn't look like `activeItem` is used in render or logic?

kmcfaul replied:
`activeItem` indicates the selected item (set by onSelect)

redallen replied:
Right, but I only see it used 3 places in the 2 drilldown examples:
```js
this.state = {
      **activeItem: null,
      menuDrilledIn: [],
      drilldownPath: [],
      menuHeights: {},
      activeMenu: 'rootMenu'
    };
 ...
 this.onSelect = (event, itemId) => {
      if (itemId.startsWith('group:')) {
        console.log(`selected sub-menu: ${itemId.split('group:')[1]}`);
      } else {
        console.log(`selected: ${itemId}`);
        this.setState({
          **activeItem: itemId
        });
      }
    };
 ...
 const { **activeItem, menuDrilledIn, drilldownPath, activeMenu, menuHeights } = this.state;
```
Look like it's not used anywhere in the render logic and can be safely removed.

kmcfaul replied:
Yeah I can remove them

redallen reviewed (changes requested):
Wow this component is getting complicated! I wonder if there's more ways than what I've mentioned to make implementing drilldown menus simpler for the consumer?

I'm thinking about the API more broadly now and I think it might be possible for the consumer to not assign `itemId`s, `onClick`s, `isOnPath`, and `isMenuDrilledIn` at all. Couldn't we do that using `React.Children` in `Menu` if it has `containsDrilldown` and use index-based `itemId`s instead? Then if a consumer wants manual control of the drilled down menu they can pass `drilldownPath=[1,2,3]` to indicate the 3rd MenuItem of the 2nd MenuItem of the 1st MenuItem is drilled in.

redallen reviewed (changes requested):
This is looking much cleaner!! I still think `activeItem` and `onSelect` can be removed. Could the `setHeight`s be moved to a different example or is it a designer necessity? I don't like how the menu changes height and potentially scrolls off the screen.

kmcfaul commented:
@mcarrano @mcoker Any thoughts about the changing height?

I think it may require a core update, I tried removing the height and it appears to make longer lists within the sub-menus truncate and un-scrollable.

tlabaj commented on packages/react-core/src/components/Menu/Menu.tsx:
```diff
@@ -26,8 +28,24 @@ export interface MenuProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'r
+  isMenuDrilledIn?: boolean;
+  /** Indicates the drilldown path of menu itemIds */
+  drilldownPath?: string[];
+  /** Indicates the menus that are drilled in */
```
I think the description is a little confusing.  May just "array of menus that are drilled into".

tlabaj commented on packages/react-core/src/components/Menu/Menu.tsx:
```diff
@@ -26,8 +28,24 @@ export interface MenuProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'r
+  /** Indicates if a menu is drilled into */
+  isMenuDrilledIn?: boolean;
+  /** Indicates the drilldown path of menu itemIds */
+  drilldownPath?: string[];
```
Maybe `drilldownPaths`?  The current name implies one path to me but the type is an array.

kmcfaul replied:
My interpretation is that it is a path of nodes. Maybe `drilldownItemPath`?

jschuler commented on packages/react-core/src/components/Menu/Menu.tsx:
```diff
@@ -58,15 +77,45 @@ class MenuBase extends React.Component<MenuProps, MenuState> {
     } = this.props;
-
+    const _isMenuDrilledIn =
+      (isMenuDrilledIn && isMenuDrilledIn) || (drilledInMenus && drilledInMenus.includes(id)) || false;
```
This looks a bit funny, should it be `isMenuDrilledIn !== undefined && isMenuDrilledIn || ...`

kmcfaul replied:
That should be equivalent

jschuler reviewed (commented):
Thanks for picking this up @kmcfaul ! Since this "variation" adds a bunch of new props you could also consider making it a dedicated component instead and not rolling the props into the Menu component.
For example you could create a new InternalMenu component that has ALL the props, but the public facing DrilldownMenu and Menu components would only surface a few of them.

Anyways, example works great!

mcarrano commented:
> @mcarrano @mcoker Any thoughts about the changing height?

Not sure what to do about this. Would it work to set a max-height after which a scrollbar is exposed? Ideas @mcoker ?

kmcfaul commented:
Do you think its a blocker or is this something we can open a follow up issue for? @mcarrano

mcarrano commented:
I think we can address the height issue as follow-up @kmcfaul .

redallen reviewed (dismissed):
In that case @mcarrano this LGTM!

tlabaj commented on packages/react-core/src/components/Menu/DrilldownMenu.tsx:
```diff
@@ -0,0 +1,45 @@
+        id={id}
+        parentMenu={menuId}
+        isMenuDrilledIn={isMenuDrilledIn}
+        {...props}
```
is there a reason we don't spread the props at the end?

jessiehuff reviewed (dismissed):
Currently there's still an [open issue for the keyboard interaction](https://github.com/patternfly/patternfly-react/issues/5144) for this component. I think there are also some refinements that could be made in the future for screen reader users (for example, there's currently nothing announcing to users if a menu item is able to be expanded so it's hard to tell when a screen reader user can continue to drill into the menu). However, I don't think that's specific to the drilldown and might exist with the flyout variations of menu too so probably doesn't need to be addressed in this PR.

tlabaj reviewed (approved):
LGTM

mcarrano reviewed (dismissed):
Looks good. I will add the keyboard interaction issue to the coming sprint and open a new issue for addressing the height.

mcoker commented on packages/react-core/src/components/Menu/examples/Menu.md:
```diff
@@ -692,3 +693,459 @@ class MenuOptionMultiSelect extends React.Component {
+              direction="down"
+              drilldownMenu={
+                <DrilldownMenu id="drilldownMenuStart">
+                  <MenuBreadcrumb>
```
In core this is constructed a bit differently. The breadcrumbs aren't a list item adjacent to the other items in the dropdown menu list. `.pf-c-menu__breadcrumb` (`<MenuBreadcrumb>`) is a direct child of `.pf-c-menu` (`<Menu>`). After that is a `<Divider>` (`<hr>`), then `.pf-c-menu__content` (`<MenuContent>`) comes after the divider. Here's a screenshot:

<img width="435" alt="Screen Shot 2021-03-02 at 5 47 53 PM" src="https://user-images.githubusercontent.com/35148959/109730773-a557f480-7b7f-11eb-9639-8d635932b6a7.png">

That said, we can update this if the markup needs to be different. I'm curious what @jessiehuff thinks. At a minimum though, if we keep the markup in this PR and update core to match, `<MenuBreadcrumb>` needs to be an `<li>` since it's a child of `<MenuList>` (a `<ul>`)

cc @mattnolting

mattnolting replied:
The back/up links follow the menus as you drillin and are relative to them, which I believe is the desired effect. 

`<MenuBreadcrumb>` is a child of `pf-c-menu`, rather than `<MenuList>`, so it should be a `<div>`. Otherwise you see a new set of breadcrumbs scroll in with each drilldown, which I don't think is the desired effect. @mcarrano @doruskova wdyt?

mcoker commented on packages/react-core/src/components/Breadcrumb/BreadcrumbItem.tsx:
```diff
@@ -48,13 +48,18 @@ export const BreadcrumbItem: React.FunctionComponent<BreadcrumbItemProps> = ({
         </span>
       )}
+      {component === 'button' && (
+        <button className={className} aria-current={ariaCurrent}>
```
this needs `type="button"`

mcoker commented:
> @mcoker I do have a question about using the link version of BreadcrumbHeading, which is in the core example. Is it ok just to pass a string for the drilldown? An href of "#" will reset the whole drilldown (because the page reloads).

@kmcfaul sorry I just saw this. Could it be a button instead like the other items? While what you have works, the parent item to the text and divider is a flex layout (or it will be once this bug is fixed - https://github.com/patternfly/patternfly-react/issues/5508), and ideally all of the item text needs to be wrapped in an element. I suppose the `component` used for the breadcrumb item link could be a `<span>`, too. 

I created https://github.com/patternfly/patternfly/issues/3896 as a follow up for whatever we implement in this PR.

kmcfaul commented:
- Added similar button setup to BreadcrumbHeading to match BreadcrumbItem
- Added `type="button"` to both
- Changed `MenuBreadcrumb` from `div` to `li`

mattnolting commented:
@kmcfaul Sorry, I'm just catching up with this one. This is looking great! `MenuBreadcrumb` isn't a child of a`<ul>` intentionally, so a new set of `Breadcrumb`s isn't scrolled in with each drill down.

mcoker reviewed (changes requested):
It looks like the last breadacrumb item is missing the `.pf-c-breadcrumb__link` class, which is causing these border styles.

<img width="333" alt="Screen Shot 2021-03-03 at 3 49 45 PM" src="https://user-images.githubusercontent.com/35148959/109877617-dd6f3e00-7c38-11eb-83ef-e0cda5957b1b.png">


Also I'm noticing I either get sent into the wrong menu or the breadcrumbs look incorrect when I'm navigating. Here's a video of going into "start rollouts", then "pause rollouts", then back to "start rollouts" but the breadcrumbs say "pause rollouts"


https://user-images.githubusercontent.com/35148959/109877630-e102c500-7c38-11eb-85b3-6ba8f4655e1a.mp4

kmcfaul commented:
@mcoker Fixed the incorrect drill out (paths weren't getting updated properly). Also fixed the styling on the BreadcrumbHeading not getting applied

tlabaj commented:
@kmcfaul if you pushed any changes last night,  I accidentally overwrote them.

mattnolting reviewed (commented):
Great work! Here are my observations about `breadcrumb` placement. 

If you think about the relationship of a back button vs breadcrumbs, the back button is relative to the current menu, so it makes sense that that button would scroll with the menu. Breadcrumbs, on the other hand, allow you to traverse all the way back to the root of the tree, so they shouldn’t be relative to individual sub-lists. Also, if the breadcrumbs were to wrap (edge case as they should become a dropdown after several level) it alters the height of the menu because the line wrap throws of the height calculation. If all the animation is isolated to the lists and sublists, the calculation is more accurate/predictable as list items truncate text.

Here's a screenshot of that with a width of 350px, which is a reasonable expected width: 
<img width="466" alt="Screen Shot 2021-03-04 at 2 42 23 PM" src="https://user-images.githubusercontent.com/5385435/110020658-d5c29e80-7cf7-11eb-969d-b181b62eb8ca.png">

kmcfaul commented:
@tlabaj Updated to resolve the conflict

mcarrano reviewed (approved):
This behaves much better with the changes to the breadcrumb. Thanks @kmcfaul @tlabaj !

mattnolting reviewed (approved):
It's perfect 🤩!! Great work, thank you!

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@4.10.12
 - @patternfly/react-code-editor@4.2.1
 - @patternfly/react-console@4.3.12
 - @patternfly/react-core@4.101.2
 - @patternfly/react-docs@5.16.16
 - @patternfly/react-inline-edit-extension@4.6.47
 - demo-app-ts@4.80.1
 - @patternfly/react-integration@4.81.1
 - @patternfly/react-table@4.23.13
 - @patternfly/react-topology@4.7.44
 - @patternfly/react-virtualized-extension@4.6.45

Thanks for your contribution! :tada:
